File: node/src/runner.ts
import type { Case, FlakegateConfig } from "./config.js";
import { getProvider } from "./providers/index.js";
import type { Provider } from "./providers/index.js";
import { ProviderError } from "./providers/index.js";
import { type ScoreResult, scoreResponses } from "./scoring.js";

export const MAX_RETRIES = 3;
export let RETRY_BACKOFF_MS = 1000;

/** Test-only hook to speed up retry-backoff tests. Not part of the public API. */
export function _setRetryBackoffMsForTests(ms: number): void {
  RETRY_BACKOFF_MS = ms;
}

export interface CaseResult {
  caseId: string;
  threshold: number;
  score: ScoreResult;
  rawResponses: string[];
  errors: string[];
  promptTokens: number;
  completionTokens: number;
}

export interface RunResult {
  passed: boolean;
  caseResults: CaseResult[];
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function callWithRetry(
  provider: Provider,
  singleCase: Case,
): Promise<{ text?: string; error?: string; promptTokens: number; completionTokens: number }> {
  let lastError: string | undefined;
  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      const result = await provider.complete(singleCase.prompt, singleCase.messages, { ...singleCase.params });
      return { text: result.text, promptTokens: result.promptTokens, completionTokens: result.completionTokens };
    } catch (err) {
      lastError = err instanceof ProviderError ? err.message : (err as Error).message;
      if (attempt < MAX_RETRIES) {
        await sleep(RETRY_BACKOFF_MS * attempt);
      }
    }
  }
  return { error: lastError, promptTokens: 0, completionTokens: 0 };
}

async function runPool<T>(tasks: (() => Promise<T>)[], concurrency: number): Promise<T[]> {
  const results: T[] = new Array(tasks.length);
  let nextIndex = 0;

  async function worker(): Promise<void> {
    while (nextIndex < tasks.length) {
      const index = nextIndex++;
      results[index] = await tasks[index]();
    }
  }

  const workers = Array.from({ length: Math.min(concurrency, tasks.length) }, () => worker());
  await Promise.all(workers);
  return results;
}

export async function runCase(provider: Provider, singleCase: Case, concurrency: number): Promise<CaseResult> {
  const tasks = Array.from({ length: singleCase.repeat }, () => () => callWithRetry(provider, singleCase));
  const outcomes = await runPool(tasks, concurrency);

  const responses: string[] = [];
  const errors: string[] = [];
  let promptTokens = 0;
  let completionTokens = 0;

  for (const outcome of outcomes) {
    if (outcome.error !== undefined) {
      errors.push(outcome.error);
    } else if (outcome.text !== undefined) {
      responses.push(outcome.text);
      promptTokens += outcome.promptTokens;
      completionTokens += outcome.completionTokens;
    }
  }

  if (responses.length === 0) {
    throw new ProviderError(
      `case '${singleCase.id}': all ${singleCase.repeat} calls failed. First error: ${errors[0] ?? "unknown"}`,
    );
  }

  const score = scoreResponses(responses, singleCase.normalize, singleCase.threshold, singleCase.extract);
  return {
    caseId: singleCase.id,
    threshold: singleCase.threshold,
    score,
    rawResponses: responses,
    errors,
    promptTokens,
    completionTokens,
  };
}

export async function runConfig(config: FlakegateConfig): Promise<RunResult> {
  const provider = getProvider(config.provider, config.model, config.apiKeyEnv);
  const caseResults: CaseResult[] = [];
  for (const singleCase of config.cases) {
    caseResults.push(await runCase(provider, singleCase, config.concurrency));
  }
  const passed = caseResults.every((r) => r.score.passed);
  return { passed, caseResults };
}

File: python/src/flakegate/runner.py
from __future__ import annotations

import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass, field

from flakegate.config import Case, FlakegateConfig
from flakegate.providers import Provider, ProviderError, get_provider
from flakegate.scoring import ScoreResult, score_responses

MAX_RETRIES = 3
RETRY_BACKOFF_SECONDS = 1.0


@dataclass(frozen=True)
class CaseResult:
    case_id: str
    threshold: float
    score: ScoreResult
    raw_responses: list[str]
    errors: list[str] = field(default_factory=list)
    prompt_tokens: int = 0
    completion_tokens: int = 0


@dataclass(frozen=True)
class RunResult:
    passed: bool
    case_results: list[CaseResult]


def _call_with_retry(provider: Provider, case: Case) -> tuple[str | None, str | None, int, int]:
    last_error: str | None = None
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            result = provider.complete(case.prompt, case.messages, dict(case.params))
            return result.text, None, result.prompt_tokens, result.completion_tokens
        except ProviderError as exc:
            last_error = str(exc)
            if attempt < MAX_RETRIES:
                time.sleep(RETRY_BACKOFF_SECONDS * attempt)
    return None, last_error, 0, 0


def run_case(provider: Provider, case: Case, concurrency: int) -> CaseResult:
    responses: list[str] = []
    errors: list[str] = []
    total_prompt_tokens = 0
    total_completion_tokens = 0

    with ThreadPoolExecutor(max_workers=min(concurrency, case.repeat)) as executor:
        futures = [executor.submit(_call_with_retry, provider, case) for _ in range(case.repeat)]
        for future in as_completed(futures):
            text, error, prompt_tokens, completion_tokens = future.result()
            if error is not None:
                errors.append(error)
            else:
                responses.append(text)
                total_prompt_tokens += prompt_tokens
                total_completion_tokens += completion_tokens

    if not responses:
        raise ProviderError(
            f"case '{case.id}': all {case.repeat} calls failed. First error: {errors[0] if errors else 'unknown'}"
        )

    score = score_responses(responses, case.normalize, case.threshold, case.extract)
    return CaseResult(
        case_id=case.id,
        threshold=case.threshold,
        score=score,
        raw_responses=responses,
        errors=errors,
        prompt_tokens=total_prompt_tokens,
        completion_tokens=total_completion_tokens,
    )


def run_config(config: FlakegateConfig) -> RunResult:
    provider = get_provider(config.provider, config.model, config.api_key_env)
    case_results = [run_case(provider, case, config.concurrency) for case in config.cases]
    passed = all(result.score.passed for result in case_results)
    return RunResult(passed=passed, case_results=case_results)
